monkey/monkey_island/cc/repository/mongo_event_repository.py

```diff
@@ -0,0 +1,71 @@
+from typing import Any, MutableMapping, Sequence, Type
+
+from pymongo import MongoClient
+
+from common.event_serializers import EVENT_TYPE_FIELD, EventSerializerRegistry
+from common.events import AbstractAgentEvent
+from common.types import AgentID
+from monkey_island.cc.repository import IEventRepository
+
+from . import RemovalError, RetrievalError, StorageError
+from .consts import MONGO_OBJECT_ID_KEY
+
+
+class MongoEventRepository(IEventRepository):
+    """A repository for storing and retrieving events in MongoDB."""
+
+    def __init__(self, mongo_client: MongoClient, serializer_registry: EventSerializerRegistry):
+        self._events_collection = mongo_client.monkey_island.events
+        self._serializers = serializer_registry
+
+    def save_event(self, event: AbstractAgentEvent):
+        try:
+            serializer = self._serializers[type(event)]
+            serialized_event = serializer.serialize(event)
+            self._events_collection.insert_one(serialized_event)
+        except Exception as err:
+            raise StorageError(err)
+
+    def get_events(self) -> Sequence[AbstractAgentEvent]:
+        try:
+            serialized_events = list(self._events_collection.find())
+            return list(map(self._deserialize, serialized_events))
+        except Exception as err:
+            raise RetrievalError(f"Error retrieving events: {err}")
+
+    def get_events_by_type(
+        self, event_type: Type[AbstractAgentEvent]
+    ) -> Sequence[AbstractAgentEvent]:
+        try:
+            serialized_events = list(
+                self._events_collection.find({EVENT_TYPE_FIELD: event_type.__name__})
+            )
+            return list(map(self._deserialize, serialized_events))
+        except Exception as err:
+            raise RetrievalError(f"Error retrieving events for type {event_type}: {err}")
+
+    def get_events_by_tag(self, tag: str) -> Sequence[AbstractAgentEvent]:
+        try:
+            serialized_events = list(self._events_collection.find({"tags": {"$in": [tag]}}))
+            return list(map(self._deserialize, serialized_events))
+        except Exception as err:
+            raise RetrievalError(f"Error retrieving events for tag {tag}: {err}")
+
+    def get_events_by_source(self, source: AgentID) -> Sequence[AbstractAgentEvent]:
+        try:
+            serialized_events = list(self._events_collection.find({"source": str(source)}))
+            return list(map(self._deserialize, serialized_events))
+        except Exception as err:
+            raise RetrievalError(f"Error retrieving events for source {source}: {err}")
+
+    def reset(self):
+        try:
+            self._events_collection.drop()
+        except Exception as err:
+            raise RemovalError(f"Error resetting the repository: {err}")
+
+    def _deserialize(self, mongo_record: MutableMapping[str, Any]) -> AbstractAgentEvent:
+        del mongo_record[MONGO_OBJECT_ID_KEY]
```

Comment:
You don't need to `del` the ID if you add `{'_id': False}` to your `find()` call. See [this](https://github.com/guardicore/monkey/blob/0b5dd3832b7591cef77a3a24cd3802c1be04fb55/monkey/monkey_island/cc/repository/mongo_machine_repository.py#L72-L79) for an example.